src/vs/workbench/contrib/terminal/browser/xterm/contextualActionAddon.ts

```diff
@@ -0,0 +1,151 @@
+	 */
+	registerCommandFinishedListener(options: ITerminalContextualActionOptions): void;
+}
+
+export interface IContextualActionAdddon extends IContextualAction {
+	onDidRequestRerunCommand: Event<ITerminalCommand>;
+}
+
+export class ContextualActionAddon extends Disposable implements ITerminalAddon, IContextualActionAdddon {
+	private readonly _onDidRequestRerunCommand = new Emitter<ITerminalCommand>();
+	readonly onDidRequestRerunCommand = this._onDidRequestRerunCommand.event;
+
+	private _terminal: Terminal | undefined;
+
+	private _currentQuickFixElement: HTMLElement | undefined;
+
+	private _decorationMarkerIds = new Set<string>();
+
+	private _commandListeners: Map<string, ITerminalContextualActionOptions[]> = new Map();
+
+	private _matchActions: MatchActions | undefined;
+
+	constructor(private readonly _capabilities: ITerminalCapabilityStore,
+		@IContextMenuService private readonly _contextMenuService: IContextMenuService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService) {
+		super();
+		const commandDetectionCapability = this._capabilities.get(TerminalCapability.CommandDetection);
+		if (commandDetectionCapability) {
+			this._registerCommandFinishedHandler();
+		} else {
+			this._capabilities.onDidAddCapability(c => {
+				if (c === TerminalCapability.CommandDetection) {
+					this._registerCommandFinishedHandler();
+				}
+			});
+		}
+	}
+	activate(terminal: Terminal): void {
+		this._terminal = terminal;
+	}
+
+	showQuickFixMenu(): void {
+		this._currentQuickFixElement?.click();
+	}
+
+	registerCommandFinishedListener(options: ITerminalContextualActionOptions): void {
+		const matcherKey = options.commandLineMatcher.toString();
+		const currentOptions = this._commandListeners.get(matcherKey) || [];
+		currentOptions.push(options);
+		this._commandListeners.set(matcherKey, currentOptions);
+	}
+
+	private _registerCommandFinishedHandler(): void {
+		const terminal = this._terminal;
+		const commandDetection = this._capabilities.get(TerminalCapability.CommandDetection);
+		if (!terminal || !commandDetection) {
+			return;
+		}
+		this._register(commandDetection.onCommandFinished(async command => {
+			this._matchActions = getMatchOptions(command, this._commandListeners);
+		}));
+		// The buffer is not ready by the time command finish
+		// is called. Add the decoration on command start using the actions, if any,
+		// from the last command
+		this._register(commandDetection.onCommandStarted(() => {
+			if (this._matchActions) {
+				this._registerContextualDecoration();
+				this._matchActions = undefined;
+			}
+		}));
+	}
+
+	private _registerContextualDecoration(): void {
+		const marker = this._terminal?.registerMarker();
+		const actions = this._matchActions;
+		if (this._terminal && 'registerDecoration' in this._terminal) {
+			const decoration = (this._terminal as any).registerDecoration({ marker, actions, layer: 'top' });
```

Comment:
This in and any shouldn't be needed?

Reply:
it is saying `Property 'registerDecoration' does not exist on type 'Terminal'.ts(2339)` when I remove that

Reply:
oh that was bc I had imported `Terminal` from `xterm-headless`